zingolib/src/wallet/data.rs

```diff
@@ -36,6 +42,199 @@ pub enum PoolNullifier {
+        Ok(())
+    })?;
+    Ok(())
+}
+fn write_checkpoints<W, Cid>(mut writer: W, checkpoints: &Vec<(Cid, Checkpoint)>) -> io::Result<()>
+where
+    W: Write,
+    Cid: Ord + std::fmt::Debug + Copy,
+    u32: From<Cid>,
+{
+    Vector::write(
+        &mut writer,
+        &checkpoints,
+        |mut w, (checkpoint_id, checkpoint)| {
+            w.write_u32::<LittleEndian>(u32::from(*checkpoint_id))?;
+            match checkpoint.tree_state() {
+                shardtree::TreeState::Empty => w.write_u8(0),
+                shardtree::TreeState::AtPosition(pos) => {
+                    w.write_u8(1)?;
+                    w.write_u64::<LittleEndian>(<u64 as From<Position>>::from(pos))
+                }
+            }?;
+            Vector::write(
+                &mut w,
+                &checkpoint.marks_removed().iter().collect::<Vec<_>>(),
+                |w, mark| w.write_u64::<LittleEndian>(<u64 as From<Position>>::from(**mark)),
+            )
+        },
+    )?;
+    Ok(())
+}
+/// Write memory-backed shardstore, represented tree.
+fn write_shardtree<H: Hashable + Clone + Eq + HashSer, C: Ord + std::fmt::Debug + Copy, W: Write>(
+    tree: &mut shardtree::ShardTree<
+        MemoryShardStore<H, C>,
+        COMMITMENT_TREE_LEVELS,
+        MAX_SHARD_LEVEL,
+    >,
+    mut writer: W,
+) -> io::Result<()>
+where
+    u32: From<C>,
+{
+    // Replace original tree with empty tree, and mutate new version into store.
+    let mut store = std::mem::replace(
+        tree,
+        shardtree::ShardTree::new(MemoryShardStore::empty(), 0),
+    )
+    .into_store();
+    macro_rules! write_with_error_handling {
+        ($writer: ident, $from: ident) => {
+            if let Err(e) = $writer(&mut writer, &$from) {
+                *tree = shardtree::ShardTree::new(store, MAX_REORG);
+                return Err(e);
+            }
+        };
+    }
+    // Write located prunable trees
+    write_with_error_handling!(write_shards, store);
+    let mut checkpoints = Vec::new();
+    store
+        .with_checkpoints(MAX_REORG, |checkpoint_id, checkpoint| {
+            checkpoints.push((*checkpoint_id, checkpoint.clone()));
+            Ok(())
+        })
+        .expect("Infallible");
+    // Write checkpoints
+    write_with_error_handling!(write_checkpoints, checkpoints);
+    let cap = store.get_cap().expect("Infallible");
+    // Write cap
+    write_with_error_handling!(write_shard, cap);
+    *tree = shardtree::ShardTree::new(store, MAX_REORG);
+    Ok(())
+}
+
+impl Default for WitnessTrees {
+    fn default() -> WitnessTrees {
+        Self {
+            witness_tree_sapling: shardtree::ShardTree::new(MemoryShardStore::empty(), MAX_REORG),
+            witness_tree_orchard: shardtree::ShardTree::new(MemoryShardStore::empty(), MAX_REORG),
+        }
+    }
+}
+
+impl WitnessTrees {
+    pub(crate) fn add_checkpoint(&mut self, height: BlockHeight) {
+        self.witness_tree_sapling.checkpoint(height).unwrap();
```

Comment:
I don't understand the details of checkpoint..  I think I'd need to write it out.

Reply:
I think we delved into this sufficiently via screenshare